From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: qdio: continue polling for buffer state ERROR
References: bnc#579111,LTC#60735
Patch-mainline: Upstream submission made

Symptom:     Inbound traffic handling may hang.
Problem:     If next buffer to check is in state ERROR, polling is
             stopped and the final check for further available
             inbound buffers disregards buffers in state ERROR.
Solution:    Include state ERROR when checking availability of more
             inbound buffers.

Acked-by: John Jolly <jjolly@suse.de>
---
 drivers/s390/cio/qdio_main.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/s390/cio/qdio_main.c
+++ b/drivers/s390/cio/qdio_main.c
@@ -531,7 +531,7 @@ static inline int qdio_inbound_q_done(st
 	qdio_siga_sync_q(q);
 	get_buf_state(q, q->first_to_check, &state, 0);
 
-	if (state == SLSB_P_INPUT_PRIMED)
+	if (state == SLSB_P_INPUT_PRIMED || state == SLSB_P_INPUT_ERROR)
 		/* more work coming */
 		return 0;
 
